server/common/fbs/fbs_coders.py

```diff
@@ -0,0 +1,225 @@
+    return (array_type, array_value)
+
+
+def deserialize_typed_array(tarr):
+    type_map = {
+        TypedFBArray.TypedFBArray.NONE: (None, None),
+        TypedFBArray.TypedFBArray.Uint32FBArray: (DenseNumericCoder, Uint32FBArray.Uint32FBArray),
+        TypedFBArray.TypedFBArray.Int32FBArray: (DenseNumericCoder, Int32FBArray.Int32FBArray),
+        TypedFBArray.TypedFBArray.Float32FBArray: (DenseNumericCoder, Float32FBArray.Float32FBArray),
+        TypedFBArray.TypedFBArray.Float64FBArray: (DenseNumericCoder, Float64FBArray.Float64FBArray),
+        TypedFBArray.TypedFBArray.JSONEncodedFBArray: (PolymorphicCoder, JSONEncodedFBArray.JSONEncodedFBArray),
+        TypedFBArray.TypedFBArray.DictEncoded8FBArray: (CategoricalCoder, DictEncoded8FBArray.DictEncoded8FBArray),
+        TypedFBArray.TypedFBArray.DictEncoded16FBArray: (CategoricalCoder, DictEncoded16FBArray.DictEncoded16FBArray),
+        TypedFBArray.TypedFBArray.DictEncoded32FBArray: (CategoricalCoder, DictEncoded32FBArray.DictEncoded32FBArray),
+        TypedFBArray.TypedFBArray.SparseFloat32FBArray: (SparseNumericCoder, SparseFloat32FBArray.SparseFloat32FBArray),
+        TypedFBArray.TypedFBArray.SparseFloat64FBArray: (SparseNumericCoder, SparseFloat64FBArray.SparseFloat64FBArray),
+    }
+
+    (u_type, u) = tarr
+    if u_type is TypedFBArray.TypedFBArray.NONE:
+        return None
+
+    Coder, TarType = type_map.get(u_type, None)
+    if TarType is None:
+        raise TypeError(f"FBS contains unknown data type: {u_type}")
+    # for decoding, we require the encoded column and the type of typed array
+    return Coder().decode_array(u, TarType)
+
+
+class DenseNumericCoder:
+    n_slots = 1
+
+    def encode_array(self, array, builder, dtype):
+        # convert pandas series to numpy array
+        if isinstance(array, pd.Series):
+            array = array.to_numpy()
+        elif sp.issparse(array):
+            array = array.A.flatten()
+        # convert to the specified dtype
+        if np.dtype(array.dtype).str != dtype:
+            array = array.astype(dtype)
+
+        vec = builder.CreateNumpyVector(array)
+        builder.StartObject(self.n_slots)
+        builder.PrependUOffsetTRelativeSlot(0, vec, 0)
+        return builder.EndObject()
+
+    def decode_array(self, u, TarType):
+        arr = TarType()
+        arr.Init(u.Bytes, u.Pos)
+        return arr.DataAsNumpy()
+
+
+class CategoricalCoder:
+    n_slots = 2
+
+    def encode_array(self, array, builder, dtype):
+        if isinstance(array, pd.Series) and array.dtype.name == "category":
+            # create the code-to-value dictionary and encode in utf-8 as a byte array
+            dictionary = np.array(bytearray(json.dumps(dict(enumerate(array.cat.categories))), "utf-8"))
+            codes = array.cat.codes.values
+
+            # this type safeguard is likely unecessary but can't hurt to include
+            if np.dtype(codes.dtype).str != dtype:
```

Comment:
I'm not opposed to this, but it may hide real errors in the encoder table.  Suggest you add an assertion that the number of codes is less than max number that can be held in this code dtype.

Eg, something like:  `assert len(codes) <= np.iinfo(dtype).max + 1`